### PR TITLE
go: Add runnables for Go

### PR DESCRIPTION
**Pull request by anilsenay**
Implemented runnables for specially for running tests for Go.

I'm grateful for your feedback because this is my first experience with Rust and Zed codebase.

Release Notes:

- Added Runnables/Tasks for:
  - Run test functions which start with "Test"
  - Run subtests
  - Run benchmark tests
  - Run main function

resim

https://github.com/zed-industries/zed/assets/1047345/ae1abd9e-3657-4322-9c28-02d0752b5ccd

### CODE REVIEW

**Comment by anilsenay** [comment]
@cla-bot check

File: crates/languages/src/go/runnables.scm
Comment (osiewicz) [inline crates/languages/src/go/runnables.scm:5]: Readability nit: could we use a separate name for the thing you're matching on and run position itself?

```suggestion
(
    (function_declaration name: (_) @run @_name
        (#match? @_name "^Test.+"))
) @go-test
```
Same goes for other queries in this file.

File: crates/languages/src/go.rs
Comment (osiewicz) [inline crates/languages/src/go.rs:433]: I think it would make sense for `build_context` function to get access to context built up so far; that way, it could inspect variables set by the captures and you wouldn't have to extract function name from current line, which in my opinion isn't ideal. You really had no better tool at your disposal though.
Tomorrow I'll wire through the changes that will allow us to simplify that function a bit.

Comment (anilsenay) [reply crates/languages/src/go.rs:433]: @osiewicz Unfortunately, I haven't found a better way to get the line. As you said, It would be better if I could just access via`build_context`. I would appreciate your assistance 🙏

**Comment by d1y** [comment]
@mrnugget  https://github.com/zed-industries/zed/pull/12110 Are these two duplicates?

**Comment by aexvir** [comment]
> @mrnugget #12110 Are these two duplicates?

they do seem to overlap, yep, although this one adds support for subtests (and benchmarks) so I hope it can be rebased and merged still
most of the tests I usually run have subtests so I find it extremely convenient to be able to run specific cases

**Comment by mrnugget** [comment]
Woops, sorry! Didn't see this one here, I'll try to help with the rebase.

**Comment by anilsenay** [comment]
> Woops, sorry! Didn't see this one here, I'll try to help with the rebase.

Ahaha its okey 🙏  I haven't rebase it to avoid overlap your code. You may review my code and change as you like while doing rebase.

**Comment by mrnugget** [comment]
Alright, I rebased and combined the code. Clicking on subtests works, but there's something fishy going on that stops them from showing up in the modal. They only show up when on first character of the line that contains `t.Run`. @osiewicz is looking into it.

**Comment by anilsenay** [comment]
> Alright, I rebased and combined the code. Clicking on subtests works, but there's something fishy going on that stops them from showing up in the modal. They only show up when on first character of the line that contains `t.Run`. @osiewicz is looking into it.

Thank you @mrnugget 🎉 

Alright, he would also look for a better way to get the selected line. Then I'll leave it to him. I'll also look at what I can do.

File: crates/languages/src/go.rs
Comment (anilsenay) [inline crates/languages/src/go.rs:492]: I think `peek()` does not return the whole line. I debugged `line` and it has not contain all line content. I tried to use `collect` and then got the whole line. Could you take a look for this issue?
```
let line = snapshot.text_for_range(start..end).collect::<String>();

let go_subtest_variable = extract_subtest_name(line.as_str())
```

Comment (mrnugget) [reply crates/languages/src/go.rs:492]: It does for me. 

```
[crates/languages/src/go.rs:492:9] &line = Some(
    "\tt.Run(\"Foobar\", func(t *testing.T) {\n",
)
[crates/languages/src/go.rs:492:9] &line = Some(
    "\t\tfmt.Println(\"subtest\")\n",
)
```

**Comment by osiewicz** [comment]
I think @mrnugget 's fix did the trick (partially), as it's no longer necessary to have the selection on the first character of the line with `Run` func. 

However, there's still another outstanding problem; this task does not show up in a modal when the cursor is not on the same line as `func Run*`. That's reasonable (we can't look at the whole text), but at the same time maybe we could do something about it. Gonna dig into it.

File: crates/languages/src/go.rs
Comment (osiewicz) [reply crates/languages/src/go.rs]: No worries at all! It's not like there was a better way either.
I've opened a PR in https://github.com/zed-industries/zed/pull/12134; once that lands,
you'll be able to capture the first parameter (that's a string) as a hidden variable, process it (stripping "Run" prefix) and then save it as the `SUBTASK_NAME`.

File: crates/languages/src/go.rs
Comment (osiewicz) [inline crates/languages/src/go.rs:465]: Ideally we should add something like `ZED_FILE_DIR` so that you don't have to do this. I'll take care of it.

**Comment by anilsenay** [comment]
> I think @mrnugget 's fix did the trick (partially), as it's no longer necessary to have the selection on the first character of the line with `Run` func.
> 
> However, there's still another outstanding problem; this task does not show up in a modal when the cursor is not on the same line as `func Run*`. That's reasonable (we can't look at the whole text), but at the same time maybe we could do something about it. Gonna dig into it.

to clearify, is this the modal problem you mention?
on a line inside subtest:
resim

On the line of subtest:
resim

**Comment by osiewicz** [comment]
Yes, that's exactly the problem. It is because we're parsing the current line to look for name of subtest, but it may actually reside on a different line.

Honestly though, that's totally not your fault. It is a systemic problem (so to speak); right now we always use your current cursor position to run task context evaluation, but here it's not gonna cut it, as your context is sourced elsewhere.
E.g: If you're in the middle of a test, then that test *is* your context.
Then there's also a matter of nesting. Even though you're in a *sub*test, we should also be grabbing the context from the `func TestSomething` as well.

**Comment by anilsenay** [comment]
> Yes, that's exactly the problem. It is because we're parsing the current line to look for name of subtest, but it may actually reside on a different line.
> 
> Honestly though, that's totally not your fault. It is a systemic problem (so to speak); right now we always use your current cursor position to run task context evaluation, but here it's not gonna cut it, as your context is sourced elsewhere. E.g: If you're in the middle of a test, then that test _is_ your context. Then there's also a matter of nesting. Even though you're in a _sub_test, we should also be grabbing the context from the `func TestSomething` as well.

Got it, thank you for explanation 🙏 
I was focusing on rendering run indicators, i wasn't aware of this modal problem until you mention it, lol. 
I guess this is beyond me but I'll try to help if I figure out something via suggestion or something.

**Comment by thnt** [comment]
Since MR #12110  has been merged, I'm reporting here:
Because  #2323 has been resolved, I expect this also work for nested folder without `go.work`

**Comment by anilsenay** [comment]
> Since MR #12110 has been merged, I'm reporting here: Because #10224 has been resolved, I expect this also works for nested folder without `go.work`
> 
> I also want to able to configure custom test args, ex:
> 
> ```json
> {
>  "languages": {
>     "Go": {
>       "test_args": ["-v", "-cover", "-failfast"]
>     }
>   }
> }
> ```

Oh right, our way to set `GO_PACKAGE_TASK_VARIABLE` is not working for submodules. I think we should fix it. 
Some solutions came in mind:
1. Finding go module name of running test(from go.mod) and run test in that directory (i think it can be done with `cwd`) as {module_name}/{the_package_name_of_test}

for example, in a working directory `/a` and `/b` is different modules which has different go.mod file. When running a test such as /a/pkg/boo_test.go, 
- `cwd` should be `/a`, 
- command should be `go test module-a/pkg -run TestXxx` or path can be used without module name: `go test ./pkg -run TestXxx`

I guess vscode-go extension doing similar like this. In addition, module names are listed in the Testing tab. So getting module name of a go file may be useful for us even later.

2. Easily, using file's absolute path instead of `GO_PACKAGE_TASK_VARIABLE` like `go test /Users/user/my-app/a/b/foo_test.go - run TestXxx`

@mrnugget @osiewicz 

--- 
I agree that configurable test args could be helpful. For example I changed test timeout argument in vscode. I can be done by injecting user's custom arg list to task's args list maybe. Do you have any opinion or future plans for this suggestion @mrnugget @osiewicz ?

**Comment by osiewicz** [comment]
First off, if the user has a task in their `tasks.json` that's bound to your tag, that's gonna take the precedence over language-provided tags, so there's already a way for users to override tasks from language.
I think this is a valid use case though; we need a way for user to configure language tasks.
I imagine that we could have a section in settings.json where an user could set their own value for a specific task variable. Then, in `build_context` you could set that task variable only if there's no user-supplied value.

**Comment by osiewicz** [comment]
Ok, I've opened last PR (https://github.com/zed-industries/zed/pull/12237) that should unblock work on this one.
I'd say we should focus on getting this merged and then - if you fancy it - we can tackle @thnt s request in a follow-up PR.

File: crates/languages/src/go/runnables.scm
Comment (osiewicz) [inline crates/languages/src/go/runnables.scm:10]: Is it safe to assume that the variable we're calling `Run` on is always going to be called `t`? I get that it's a convention, but is there some precedent in Go community for breaking it sometimes?

File: crates/languages/src/go/runnables.scm
Comment (osiewicz) [inline crates/languages/src/go/runnables.scm:11]:
```suggestion
        (#match? @_name "^t.Run.*"))
        (#set! tag go-subtest)
) @go-subtest
```
You'll have to fix up the way runnables are tagged in this PR once https://github.com/zed-industries/zed/pull/12237 lands.

File: crates/languages/src/go/runnables.scm
Comment (osiewicz) [inline crates/languages/src/go/runnables.scm:23]:
```suggestion
        (#match? @_name "^main$"))
        (#set! tag go-run)
)
```

File: crates/languages/src/go/runnables.scm
Comment (osiewicz) [reply crates/languages/src/go/runnables.scm:10]: Ok, I think we can rework this pattern altogether to simplify it a bit:
With the following query:
```

; `t.Run`
(
    (call_expression 
      function: (
      	selector_expression
      	field: _ @run @_name
        (#eq? @_name "Run")
      )
      arguments: (
      	argument_list
        .
        (interpreted_string_literal ) @_subtest_name
       )
    )
    (#set! tag go-subtest)
)
```
In `build_context` the task_variables argument is gonna have the value of first argument (string literal) in `_subtest_name` key. You won't have to parse the line then. 
You still need to escape it (I guess?) and set it in the context being returned from build_context (captures with _ prefix are removed after a call to build_context so that end users can't reference it).

**Comment by anilsenay** [comment]
> Ok, I've opened last PR (#12237) that should unblock work on this one. I'd say we should focus on getting this merged and then - if you fancy it - we can tackle @thnt s request in a follow-up PR.

If you talked about submodule issue, I will open an issue for that not to be forgotten after this PR merged.

File: crates/languages/src/go/runnables.scm
Comment (kisunji) [reply crates/languages/src/go/runnables.scm:11]: Chipping in to say it's a fairly strong convention and good starting point. I've seen some codebases use test helpers that break the `t.Run` pattern but VSCode and GoLand don't match those subtests either.

Comment (osiewicz) [reply crates/languages/src/go/runnables.scm:11]: Ah, that's fair. With that other query (https://github.com/zed-industries/zed/pull/12003#discussion_r1613415317) we don't even need to match on variable name though, as we'll capture just the name of the function being called; so the whole dispute about whether we should match `t.` is kinda moot. ;p

Comment (kisunji) [reply crates/languages/src/go/runnables.scm:11]: Would your example above match an unrelated `Run()` call like
```go
client.Run(ctx) // should not match

t.Run("client returns OK", func(t testing.T) {...}) // should match
```

Comment (osiewicz) [reply crates/languages/src/go/runnables.scm:11]: No, it would not. We also expect the first argument to be a string literal. We could perhaps constrain it further to also require a second parameter that's (I guess?) an anonymous function taking a single param of type `testing T`

File: crates/languages/src/go.rs
Comment (osiewicz) [reply crates/languages/src/go.rs:465]: Ok, we'll have ZED_DIRNAME on main real soon: https://github.com/zed-industries/zed/pull/12245 ;  you won't have to set that variable then.

File: crates/languages/src/go/runnables.scm
Comment (anilsenay) [inline crates/languages/src/go/runnables.scm:36]: I changed here to check second parameter based on your mentions, could you check here @osiewicz please?

Comment (osiewicz) [reply crates/languages/src/go/runnables.scm:39]: This looks good to me! We can hide the `second_argument` from user though, but that's relatively minor:
```suggestion
        (func_literal
          parameters: (
            parameter_list
            (parameter_declaration
              name: (identifier) @_param_name
              type: (pointer_type
                (qualified_type
                  package: (package_identifier) @_pkg
                  name: (type_identifier) @_type
                  (#eq? @_pkg "testing")
                  (#eq? @_type "T")
                )
              )
            )
          )
        ) @_second_argument
```

**Review by osiewicz — APPROVED** [review]
LGTM :)

**Pull request merged**
